### PR TITLE
Entity class: replaced `getParent()` with `getParentModel()` and `getParentComponent()` (#320)

### PR DESCRIPTION
**Pull request by agarny**
Addresses issue #320.

### CODE REVIEW

**Comment by agarny** [comment]
At this point, this PR includes:

- Replacement of `Entity::getParent()` with `Entity::getParentModel()` and `Entity::getParentComponent()`.
- A small fix for one of the `ComponentEntity::takeComponent()` methods.
- ~~Fixed one our Connection tests following our use of smart pointers rather than raw pointers.~~ (see commit 1860300)

What might remain to be done:
- Rename `Entity::hasParent()` to `Entity::hasParentComponent()`?
- Remove `Entity::setParent()` and `Entity::clearParent()` from the API? (This would require `Model`, `Component` and `ComponentEntity` to be friends with `Entity`...? I am not in favour of friend classes, but is there another solution?)

**Comment by agarny** [comment]
@hsorby, Buildbot probably ought to check that the `test_clang_format` target exists (for the Clang Format test)?...

**Comment by agarny** [comment]
I have done a bit more thinking with regards to getting rid of `Entity:: setParent()`and `Entity:: clearParent()`, and this could be achieved in a nice way by improving out Pimpl approach (see issue #279).

I mean, `Entity:: setParent()`and `Entity:: clearParent()` clearly have nothing to do in the API. Yet, using our current Pimpl approach, there is no choice but to have them. However, if our Pimpl approach was to use inheritance, we wouldn't need those methods anymore.

So, I am going to implement inheritance in our Pimpl in this PR... (Ideally, I would do this in issue #279, but we already have far too many PRs waiting to be merged in and my libCellML work is getting unnecessarily complicated as a result, so...)

**Comment by agarny** [comment]
> However, if our Pimpl approach was to use inheritance, we wouldn't need those methods anymore.

Scrap that, that's not the case. :(

**Comment by agarny** [comment]
Ok, I am going to leave things as they are... for now. There are several things that I don't like / think are wrong:
- `Variable` indirectly inherits from `Entity`, which means that it could have a `Model` as a parent...!?
- Conversely, because `Units` also inherits from `Entity`, a `Units` could have a `Component` as a parent...!? (It would have been fine in CellML 1.0/1.1, but not in CellML 2.0.)
- I would rename our two `Entity::setParent()` to `Entity::setParentModel()` and `Entity::setParentComponent()`.
- Likewise, I really think that `Entity::hasParent()` ought to be renamed `Entity::hasParentComponent` since this is really what we check.
- Etc.

Bottom line: it would be nice to have this PR merged in... after we have got the Clang Format test disabled when the `test_clang_format` target doesn't exit.

**Comment by agarny** [comment]
FWIW, this PR is ready to be merged in...

**Comment by agarny** [comment]
Ping @nickerso @hsorby

**Comment by hsorby** [comment]
Your request timed out.

**Comment by agarny** [comment]
> Your request timed out.

??

**Comment by agarny** [comment]
@nickerso and @hsorby: coverage is working once again (thanks @hsorby!), so you are (very) kindly invited to review and merge this PR in... :)

**Review by nickerso — DISMISSED** [review]

**Comment by agarny** [comment]
Ping @hsorby.

**Comment by agarny** [comment]
@hsorby, @nickerso and @kerimoyle... feel free to (re)review this PR... :) (@kerimoyle, I can't put you as a reviewer for now, but feel free to add yourself once you have agreed to be part of the "team".)

**Review by hsorby — COMMENTED** [review]
With all the changes from using `libcellml::Model` to `libcellml::ModelPtr` and the like do we still have the former?  I think we still need to test these ways of creating libCellML objects.

**Comment by agarny** [comment]
Yes, [`coverage/coverage.cpp`](https://github.com/agarny/libcellml/blob/issue320/tests/coverage/coverage.cpp), [`model/units_import.cpp`](https://github.com/agarny/libcellml/blob/issue320/tests/model/units_import.cpp) and [`units/units.cpp`](https://github.com/agarny/libcellml/blob/issue320/tests/units/units.cpp) still make use of `libcellml::Model`.

**Comment by agarny** [comment]
Ok, I have just tested this PR against LLVM coverage and I am not getting 100% (unlike the `develop` branch), even though GCC coverage says everything is fine. So, I am going to fix that...

**Comment by agarny** [comment]
Ok, back to normal! (Oopsie!)

**Comment by kerimoyle** [comment]
Hi Alan
I'm just going through this PR and had a few thoughts ... 

> Variable indirectly inherits from Entity, which means that it could have a Model as a parent...!?
> Conversely, because Units also inherits from Entity, a Units could have a Component as a parent...!? (It would have been fine in CellML 1.0/1.1, but not in CellML 2.0.)

I find the use of an inheritance structure overall a bit contrived - the entities themselves do not so much "inherit" the attributes of their parent, the only have a connection to the parent which (in my head anyway) isn't really the same thing.  So it's to be expected that things like variables with model parents and units with component parents exist - that's part of the inheritance structure we have.  With that in mind, I think that it would make it better that the functions were named more clearly, as you suggest below, even though they're doing the same thing under the hood (ie: setParentModel is doing the same thing as setParentComponent, just pointing to somewhere else).

> I would rename our two Entity::setParent() to Entity::setParentModel() and Entity::setParentComponent().

I think that's a good idea ... it was the first thing I went looking for and would make reading the code cleaner.

> Likewise, I really think that Entity::hasParent() ought to be renamed Entity::hasParentComponent since this is really what we check.

As above.

**Comment by agarny** [comment]
I think we will, at some point, have to review our object diagram. I will have a think about it when I am back from my soon-to-start holiday in early September... :)

**Comment by agarny** [comment]
Ping @hsorby, @nickerso & @kerimoyle.

**Comment by kerimoyle** [comment]
Hi @agarny.  I've had a look, and as in earlier comments think this is a great idea.  I wonder if the same could be extended to the other ambiguous functions?  eg:

```
bool Entity::hasParent(const ComponentPtr &component) const
{
    bool hasParent = false;
    ComponentPtr parentComponent = mPimpl->mParentComponent.lock();
    if (parentComponent == component) {
        hasParent = true;
    } else if (parentComponent) {
        hasParent = parentComponent->hasParent(component);
    }
    return hasParent;
}
```
Do we need to rather call it it `hasParentComponent` and create a similar sister function for the `model` parent?  Or just overload with a different `hasParent` argument of `model`?

**Review by nickerso — APPROVED** [review]

**Comment by nickerso** [comment]
> .. Do we need to rather call it it `hasParentComponent` ...

@agarny's comment on #320 state exactly that :) but this can be done as a separate issue/pr in future. Need to keep the focus on addressing problems with the minimal changes needed so we can get to the 1.0 release.

**Review by kerimoyle — APPROVED** [review]
Looks fine to me if we're happy to make things more consistent (as in comments) later on.

**Comment by agarny** [comment]
Two people approved, so one of you can feel free to merge this in... :)

**Comment by kerimoyle** [comment]
Ah, sorry, I was leaving you the honours :) All done now.

**Comment by agarny** [comment]
> Ah, sorry, I was leaving you the honours :) All done now.

Very kind of you, but the author of a PR can't merge his/her own work. Would be too easy otherwise... :)
